src/opening-period/form/OpeningPeriodForm.tsx

```diff
@@ -0,0 +1,315 @@
+          : '',
+        endTime: apiTimeSpan.end_time
+          ? dropMilliseconds(apiTimeSpan.end_time)
+          : '',
+        resourceState: apiTimeSpan.resource_state || ResourceState.OPEN,
+        weekdays: weekdays as FormWeekdays,
+      };
+    });
+}
+
+const formatResourcePeriodOptionsToFormFormat = (
+  resourceStateApiOptions: ResourceStateApiOption[]
+): ResourceStateOption[] =>
+  resourceStateApiOptions.map((optionInApiFormat: ResourceStateApiOption): {
+    value: string;
+    label: string;
+  } => {
+    return {
+      value: optionInApiFormat.value,
+      label: `${
+        typeof optionInApiFormat.display_name === 'string'
+          ? optionInApiFormat.display_name
+          : optionInApiFormat.display_name.fi
+      }`,
+    };
+  });
+
+export default function OpeningPeriodForm({
+  formId,
+  datePeriod,
+  resourceId,
+  datePeriodOptions,
+  submitFn,
+  successTextAndLabel,
+  errorTextAndLabel,
+}: {
+  formId: string;
+  datePeriod?: DatePeriod;
+  resourceId: number;
+  datePeriodOptions: DatePeriodOptions;
+  submitFn: (datePeriod: DatePeriod) => Promise<DatePeriod>;
+  successTextAndLabel: NotificationTexts;
+  errorTextAndLabel: NotificationTexts;
+}): JSX.Element {
+  const language = Language.FI;
+
+  const resourceStateOptionsInApiFormat: ResourceStateApiOption[] =
+    datePeriodOptions.actions.POST.resource_state.choices;
+
+  const resourceStateOptions: ResourceStateOption[] = formatResourcePeriodOptionsToFormFormat(
+    resourceStateOptionsInApiFormat
+  );
+
+  const spans: TimeSpanFormFormat[] | {}[] =
+    datePeriod?.time_span_groups[0] &&
+    datePeriod?.time_span_groups[0].time_spans
+      ? formatApiTimeSpansToFormFormat(
+          datePeriod?.time_span_groups[0].time_spans
+        )
+      : [{}];
+
+  const [periodBeginDate, setPeriodBeginDate] = useState<Date | null>(
+    datePeriod?.start_date ? new Date(datePeriod?.start_date) : null
+  );
+  const [periodEndDate, setPeriodEndDate] = useState<Date | null>(
+    datePeriod?.end_date ? new Date(datePeriod?.end_date) : null
+  );
+
+  const formValues: OpeningPeriodFormData = {
+    openingPeriodTitle: datePeriod?.name[language] || '',
+    openingPeriodOptionalDescription: datePeriod?.description[language] || '',
+    openingPeriodBeginDate: datePeriod?.start_date,
+    openingPeriodEndDate: datePeriod?.end_date,
+    timeSpans: spans,
+  };
+
+  const { register, handleSubmit, errors, control, setValue } = useForm<
+    OpeningPeriodFormData
+  >({
+    mode: 'all',
+    defaultValues: formValues,
+  });
+  const { fields, append, remove } = useFieldArray({
+    control,
+    keyName: 'timeSpanUiId',
+    name: 'timeSpans',
+  });
+  const history = useHistory();
+  const [submitStatus, setSubmitStatus] = useState<SubmitStatus>('init');
+
+  const onSubmit = async (data: OpeningPeriodFormData): Promise<void> => {
+    const validTimeSpans: TimeSpanFormFormat[] = data.timeSpans.filter(
```

Comment:
Is it possible for the user to input timeSpans that are not valid? Does this filter out those timeSpans that were added but where no field was inputted?

Reply:
I do not think so. But to change this requires that we need to come up another way to initialise empty timespan

Reply:
I tested this and validation will catch invalid timespans